src/udev/udev-builtin-blkid.c

```diff
@@ -138,13 +138,49 @@ static int find_gpt_root(UdevEvent *event, blkid_probe pr) {
         assert(pr);
 
         /* Iterate through the partitions on this disk, and see if the UEFI ESP or XBOOTLDR partition we
-         * booted from is on it. If so, find the first root disk, and add a property indicating its partition
-         * UUID. */
+         * booted from is on it. If so, find the newest root partition, and add a property indicating its
+         * partition UUID. We also do this if we are dealing with a loopback block device whose "backing
+         * filename" field is set to the string "root". In the latter case we do not search for ESP or
+         * XBOOTLDR. */
+
+        const char *devtype;
+        r = sd_device_get_devtype(dev, &devtype);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to query devtype from device: %m");
+
+        if (!streq(devtype, "disk")) {
+                log_device_debug_errno(dev, r, "Skipping GPT root logic on partition block device.");
+                return 0;
+        }
+
+        sd_id128_t esp_or_xbootldr = SD_ID128_NULL;
+        bool need_esp_or_xbootldr;
```

Comment:
These are used across the function, so it feels weird to define it here. Plus, found_esp_or_xbootldr and root_id are defined at the beginning...